db.go

```diff
@@ -16,73 +16,87 @@ var (
 	// This should create the permit if it does not exist. A call to this function is
 	// followed by a call to SessionDB.PermitChanged. identity is canonicalized before being stored
-	SetPermit(identity string, permit *Permit) error
-	RenameIdentity(oldIdent, newIdent string) error // Rename an identity. Returns ErrIdentityPermitNotFound if oldIdent does not exist. Returns ErrIdentityExists if newIdent already exists.
-	Close()                                         // Typically used to close a database handle
+	SetPermit(userId UserId, permit *Permit) error
+	Close() // Typically used to close a database handle
 }
 
 // A Session database is essentially a key/value store where the keys are
 // session tokens, and the values are tuples of (Identity,Permit).
 // All operations except for Close must be thread-safe.
 type SessionDB interface {
-	Write(sessionkey string, token *Token) error         // Set a token
-	Read(sessionkey string) (*Token, error)              // Fetch a token
-	Delete(sessionkey string) error                      // Delete a token (used to implement "logout")
-	PermitChanged(identity string, permit *Permit) error // Assign the new permit to all of the sessions belonging to 'identity'
-	InvalidateSessionsForIdentity(identity string) error // Delete all sessions belonging to the given identity. This is called after a password has been changed, or an identity renamed.
-	Close()                                              // Typically used to close a database handle
+	Write(sessionkey string, token *Token) error       // Set a token
+	Read(sessionkey string) (*Token, error)            // Fetch a token
+	Delete(sessionkey string) error                    // Delete a token (used to implement "logout")
+	PermitChanged(userId UserId, permit *Permit) error // Assign the new permit to all of the sessions belonging to 'identity'
+	InvalidateSessionsForIdentity(userId UserId) error // Delete all sessions belonging to the given identity. This is called after a password has been changed, or an identity renamed.
+	Close()                                            // Typically used to close a database handle
 }
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
 // Authenticator that simply stores identity/passwords in memory
 type dummyAuthenticator struct {
-	users     map[string]*dummyUser
+	users     map[UserId]*dummyUser
+	usersLock sync.RWMutex
+}
+
+type dummyUserStore struct {
+	users     map[UserId]*dummyUser
 	usersLock sync.RWMutex
 }
 
 type dummyUser struct {
+	userId             UserId
 	name               string
 	password           string
 	passwordResetToken string
 }
 
-func newDummyAuthenticator() *dummyAuthenticator {
+func newDummyAuthenticator(dummyUserStore *dummyUserStore) *dummyAuthenticator {
 	d := &dummyAuthenticator{}
-	d.users = make(map[string]*dummyUser)
+	d.users = dummyUserStore.users
```

Comment:
This is a big no-no in maintaining sanity when writing multithreaded code. What's happening here is that you're disassociating the lock (dummyUserStore.usersLock) from the thing that it locks (dummyUserStore.users). That's opening up the door to bugs, because it's no longer clear what lock you need to take when reading or writing to 'users'.
What you should do instead here, is to make just a single struct, called dummyUserStoreAndAuth, which will act both as an authenticator, as well as a user store.